### ISSUE TITLE
Better tab-completed navigation methods

### ISSUE THREAD

**Issue by giovannipizzi** [aiidateam/aiida_core#2230]
Currently `node.inp.<LABEL>` returns the incoming nodes of `node` with label `<LABEL>`. Similarly, `node.out.<LABEL>` for outgoing nodes. These were implemented when only `input` and `create` links existed - with this limitation, for `.inp` the result was unique given a label, while for `.out` multiple possibilities were possible (links from data to calc) so beside `node.out.<LABEL>` (pointing to the *oldest* link with that label) also a number of `node.out.<LABEL>_<PK>` were presented.

The advantage of these two methods is to be able to very quickly browse the graph with TAB-completion.

## Current issues
- we now have extended the link types: we would need something able to distinguish between them
- for incoming links, it is not anymore true that `node.inp.<LABEL>` is unique given a label (think to links from WorkflowNodes to a Data node). So only one (random) link is now shown.

## Considerations
While `node.get_incoming` and `node.get_outgoing` provide all functionality needed, they are not TAB-completable, that is still a feature we would like to preserve (not as the default method of browsing the graph, but as a convenience in the shell).

## Notes on TAB-completion
By default, in ipython only attributes are completed, as well as keys of dictionaries. However, after a string like `d['key'].` no TAB-completion is performed, and neither after a function call `d().`.

**Note**: This behavior can be changed by activating the greedy mode of ipython, by running in the current interpreter
```ipython
%config IPCompleter.greedy = True
```
While interesting (as this could allow more powerful TAB-completion), this shouldn't be suggested to the users I believe, as any function would be TAB completed, and some of them might have side effects (imagine something like `object.delete().<TAB>`, actually deleting the object).

We propose below a solution (sub-ideal, but still better than the current version) still working inside ipython.
To lift some of the limitations, another option (that we are not going to implement for now, but could be the topic of another issue) is to implement TAB-completion outside ipython (e.g. in a "virtual shell" that we need to create, or e.g. in bash).

## Suggestion
- For a ProcessNode, inputs and outputs (create or return) are unique given the label. Then we can have
  `process.inp.<LABEL>` and `process.out.<LABEL>` working as before, but limited to `input_calc/input_work` and `create/return`.
- For Data nodes, creators are unique: we can simply define a property `data.creator` to return the creator, if any (to decide if this returns None or raises if no creator is defined - I'd favour `None` but TBD).
- For "returners" of Data, and input links (both input_work and input_calc) going out Data, we don't have these constraints. Then (unless a better solution is proposed) we keep a logic like the old one. Suggested names, TBD:
  - `data.to_calc_as.<LABEL>_<PK>.<TAB>` (follow outgoing `input_calc`)
  - `data.to_work_as.<LABEL>_<PK>.<TAB>` (follow outgoing `input_work`)
  - `data.from_work_as.<LABEL>_<PK>.<TAB>` (follow ingoing `return`)

**Comment by giovannipizzi** [comment]
As this is a backward-incompatible change we should probably implement this in 1.0.0b1

**Comment by ltalirz** [comment]
As promised, here is my opinion:

 1. Our API already contains a large number of functions that users have to learn, and I think we should add new ones only when it is really necessary. Although the link names might make things a bit awkward; I think this does not warrant introducing 3/4 new functions. I prefer the solution of keeping only one attribute for input links and one for output links

 1. I think it is not a bad idea to always append the PK number in order to discourage people from using it in scripts. If there is only one, it will anyhow autocomplete.

 1. Since the corresponding functions are now named `get_incoming` and `get_outgoing`, the natural name for the inp/out would seem to be `incoming` and `outgoing` (tab completion also means one doesn't have to type the full string). Otherwise, one has to learn the inp <> get_incoming, out <> get_outgoing relation, which is not obvious.

**Comment by sphuber** [comment]
Current proposal is as follows:
```
Calculation -> Data         CREATE     |  outputs.<label>       |  creator         |
Workflow -> Data            RETURN     |  outputs.<label>       |                  |
Data -> Calculation         INPUT_CALC |                        |  inputs.<label>  |
Data -> Workflow            INPUT_WORK |                        |  inputs.<label>  |
Workflow -> Calculation     CALL_CALC  |  *called.<label>_<pk>  |  caller          |
Workflow -> Workflow        CALL_WORK  |  *called.<label>_<pk>  |  caller          |
```
The methods that would start from a `Data` node, with the exception of `creator` which is always unique, turn out to be too complex and will be difficult to understand. In this case it will be better for users to resort to `get_incoming/get_outgoing`. The links for going down a `CALL` link are marked with an asterisk as we think they might become useful, especially after #2559 is implemented and users can specify the labels of `CALL` links. However, the naming might require some more discussion

**Comment by ltalirz** [comment]
Just for the record, here the example that @giovannipizzi  gave for why a generic navigation method based on link labels is confusing in AiiDA (and thus why a generic `.ingoing/.outgoing` property would be difficult to use).

The problem is that links between Data and Calculation nodes are often named after the Data, irrespective of their directionality (this simply comes from the graph topology - calculations tend to have many more links than data).
E.g. `calc.outgoing.structure.outgoing.structure` would actually refer to a calculation node.

So, this is a consequence of the liberal regime on link labels in AiiDA - in the open provenance model, for example, the label of a link "A -> B" must form a sentence where A is the subject and B is the object (e.g. 'wasGenereratedBy', 'used', 'contained').

**Comment by giovannipizzi** [comment]
Note that I changed @sphuber's [comment](https://github.com/aiidateam/aiida_core/issues/2230#issuecomment-469285211) above to replace `creator -> created_by` and `caller -> called_by` as these are actually already implemented.

**Comment by sphuber** [comment]
@giovannipizzi I would actually keep the original proposal and just change the name of the current implementation. I think without the `_by` suffix is cleaner

**Comment by giovannipizzi** [comment]
Ok reverted back

**Comment by sphuber** [comment]
Fixed in PR #2569
